archipelago/pipeline.py

```diff
@@ -0,0 +1,1028 @@
+
+    for tile_id, place in placement.items():
+        if (x,y) == place:
+            if reg:
+                if tile_id[0] == 'r':
+                    return tile_id
+            elif y == 0:
+                assert tile_id[0] == "I" or tile_id[0] == "i"
+                if tile_id[0] == "I" and bw == 16:
+                    return tile_id
+                elif tile_id[0] == "i" and bw == 1:
+                    return tile_id
+            elif (x+1) % 4 == 0:
+                assert tile_id[0] == "m"
+                return tile_id
+            else:
+                assert tile_id[0] == "M" or tile_id[0] == "p", tile_id
+                if port in pond_ports:
+                    if tile_id[0] == "M":
+                        return tile_id
+                else:         
+                    if tile_id[0] != "M":             
+                        return tile_id      
+
+    return None
+
+
+def construct_graph(placement, routes, id_to_name):
+    graph = pnr_graph.Graph()
+    graph.id_to_name = id_to_name
+    max_reg_id = 0
+
+    for blk_id, place in placement.items():
+        if len(graph.id_to_name[blk_id].split("$")) > 0:
+            kernel = graph.id_to_name[blk_id].split("$")[0]
+        else:
+            kernel = None
+        node = pnr_graph.Node("tile", place[0], place[1], tile_id=blk_id, kernel = kernel)
+        graph.add_node(node)
+        max_reg_id = max(max_reg_id, int(blk_id[1:]))
+    graph.added_regs = max_reg_id + 1
+
+    for net_id, net in routes.items():
+
+        for route in net:
+            for seg1, seg2 in zip(route, route[1:]):
+                node1 = segment_to_node(seg1, net_id)
+                graph.add_node(node1)
+                node2 = segment_to_node(seg2, net_id)
+                graph.add_node(node2)
+                graph.add_edge(node1, node2)
+                
+                if node1.route_type == "PORT":
+                    tile_id = get_tile_at(node1.x, node1.y, node1.bit_width, placement, node1.port)
+                    graph.add_edge(tile_id, node1)
+                elif node1.route_type == "REG":
+                    tile_id = get_tile_at(node1.x, node1.y, node1.bit_width, placement, reg = True)
+                    graph.add_edge(tile_id, node1)
+
+                if node2.route_type == "PORT":
+                    tile_id = get_tile_at(node2.x, node2.y, node2.bit_width, placement, node2.port)
+                    if tile_id[0] == "m" and not "chain" in node2.port:
+                         node2.reg = True
+                    graph.add_edge(node2, tile_id)
+                elif node2.route_type == "REG":
+                    tile_id = get_tile_at(node2.x, node2.y, node2.bit_width, placement, reg = True)
+                    node2.reg = True
+                    graph.add_edge(node2, tile_id)
+
+    
+    graph.update_sources_and_sinks()
+    graph.update_edge_kernels()
+    if 'PIPELINED' in os.environ and os.environ['PIPELINED'] == '1':

```

Comment:
Consider making it as a function argument and detect it in a centralized places.